core/src/main/java/org/elasticsearch/common/geo/builders/GeometryCollectionBuilder.java

```diff
@@ -132,4 +130,39 @@ public Shape build() {
         //note: ShapeCollection is probably faster than a Multi* geom.
     }
 
+    @Override
+    public int hashCode() {
+        return Objects.hash(shapes);
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+        if (this == obj) {
+            return true;
+        }
+        if (obj == null || getClass() != obj.getClass()) {
+            return false;
+        }
+        GeometryCollectionBuilder other = (GeometryCollectionBuilder) obj;
+        return Objects.equals(shapes, other.shapes);
+    }
+
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        out.writeVInt(shapes.size());
+        for (ShapeBuilder shape : shapes) {
+            out.writeShape(shape);
+        }
+    }
+
+    @Override
+    public GeometryCollectionBuilder readFrom(StreamInput in) throws IOException {
+        GeometryCollectionBuilder geometryCollectionBuilder = new GeometryCollectionBuilder();
+        int shapes = in.readVInt();
```

Comment:
Agreed this might be a weird edge case. The parser however accepts a geometry collection with `geometries` field but no shapes, like
```
{
  "type" : "geometrycollection",
  "geometries" : [ ]
}
```
That case is also included in the random setup of the new tests. This might be a useless shape, but we have similar cases in the queries where we allow "empty" queries as edge cases. So unless it blows up later when building the geometries on the shard, I'd opt for keeping it as an edge case, wdyt? Will check if theres an integration test for this and what happens, I'd expect 0 documents in result for e.g. a GeoShapeQuery